### PR TITLE
[SecurityBundle] Fix valid provider considered undefined

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

Silly mistake.

### CODE REVIEW

**Comment by mvrhov** [comment]
This still doesn't work.
The content of array is:
```
array(1) {
  [0] =>
  string(32) "account.security.user_provider"
}
```

but `$userProvider` contains `security.user.provider.concrete.account`

**Comment by chalasr** [comment]
@mvrhov That should be better now. Could you please retry?

**Comment by mvrhov** [comment]
This seem to work now

**Comment by chalasr** [comment]
build failures unrelated. Ready

**Comment by nicolas-grekas** [comment]
any test case?

**Comment by chalasr** [comment]
now with tests :)

**Comment by jrysig** [comment]
This patch works fine for me, thank you.
With current Symfony 3.4.x-dev I'm getting error:
Invalid firewall "main": user provider "fos_userbundle" not found.

**Comment by chalasr** [comment]
Rebased

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 1]
```diff
@@ -331,10 +331,10 @@ private function createFirewall(ContainerBuilder $container, $id, $firewall, &$a
 
         // Provider id (take the first registered provider if none defined)
         if (isset($firewall['provider'])) {
-            $defaultProvider = $this->getUserProviderId($firewall['provider']);
-            if (!in_array($defaultProvider, $providerIds, true)) {
+            if (!$this->isValidProviderName($firewall['provider'], $providerIds)) {
                 throw new InvalidConfigurationException(sprintf('Invalid firewall "%s": user provider "%s" not found.', $id, $firewall['provider']));
             }
+            $defaultProvider = $providerIds[$firewall['provider']];
```
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: shouldnt the `str_replace` conversion happen here as well then?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 2]
```diff
@@ -491,10 +491,10 @@ private function createAuthenticationListeners($container, $id, $firewall, &$aut
 
                 if (isset($firewall[$key])) {
                     if (isset($firewall[$key]['provider'])) {
-                        if (!in_array($firewall[$key]['provider'], $providerIds, true)) {
+                        if (!$this->isValidProviderName($firewall[$key]['provider'], $providerIds)) {
                             throw new InvalidConfigurationException(sprintf('Invalid firewall "%s": user provider "%s" not found.', $id, $firewall[$key]['provider']));
                         }
-                        $userProvider = $this->getUserProviderId($firewall[$key]['provider']);
+                        $userProvider = $providerIds[$firewall[$key]['provider']];
```
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: same

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 3]
```diff
@@ -596,7 +596,7 @@ private function createUserProviders($config, ContainerBuilder $container)
         $providerIds = array();
         foreach ($config['providers'] as $name => $provider) {
             $id = $this->createUserDaoProvider($name, $provider, $container);
-            $providerIds[] = $id;
+            $providerIds[$name] = $id;
```
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: or normalize here :)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 4]
```diff
@@ -647,6 +647,11 @@ private function getUserProviderId($name)
         return 'security.user.provider.concrete.'.strtolower($name);
     }
 
+    private function isValidProviderName($providerName, array $allowedProviders)
```
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: perhaps make this a getter.

Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: finally removed

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 3 above)
Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: good catch, not sure about handling duplicates now

**Review by ro0NL — APPROVED** [review]

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 5]
```diff
@@ -596,7 +596,7 @@ private function createUserProviders($config, ContainerBuilder $container)
         $providerIds = array();
         foreach ($config['providers'] as $name => $provider) {
             $id = $this->createUserDaoProvider($name, $provider, $container);
-            $providerIds[] = $id;
+            $providerIds[str_replace('-', '_', $name)] = $id;
```
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: by now i think the conversion should happen in `getUserProviderId` actually :thinking: and simply make the check below `$userProvider`. That would be more robust.

```php
$userProvider = $this->getUserProviderId($firewall[$key]['provider']);
if (!isset($providerIds[$userProvider])) {
   // throw
}
```

sorry :)

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: not sure about BC :)

Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: point is we create a different id 2 lines above by passing `$name` as is to `createUserDaoProvider`.. that looks weird.

Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: I get your point but yes, changing this is not fine regarding BC. Note also that when using a custom provider (specifying `my_provider: { id: some_service }`), `createUserDaoProvider` returns the `some_service` id, it doesn't rely on `getUserProviderId`. Let's avoid any behavior change here, not the right moment for that :) 
I'm going to review this part of the extension in 2.7 once this merged, I believe there are some bugs. Right now, 3.4's security setup is broken, this covers all edge cases and makes it work again. I would like to quickly move on and polish afterwards if needed.
Btw, thanks for your review, it was useful actually.

**Comment by mvrhov** [comment]
IMO this should get merged ASAP as 3.4 is broken ATM. And then as @chalasr said it can be polished.

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**